src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1915,20 +1917,30 @@ def client_side_proxy(cmd,
                                     summary='Unable to run client proxy executable')
             raise CLIInternalError("Failed to start proxy process." + str(e))
 
-        if user_type == 'user':
-            identity_data = {}
-            identity_data['refreshToken'] = creds
-            identity_uri = f'https://localhost:{api_server_port}/identity/rt'
-
-            # Needed to prevent skip tls warning from printing to the console
-            original_stderr = sys.stderr
-            f = open(os.devnull, 'w')
-            sys.stderr = f
-
-            make_api_call_with_retries(identity_uri, identity_data, False, consts.Post_RefreshToken_Fault_Type,
-                                       'Unable to post refresh token details to clientproxy',
-                                       "Failed to pass refresh token details to proxy.", clientproxy_process)
-            sys.stderr = original_stderr
+        if not utils.is_cli_using_msal_auth():  # refresh token approach if cli is using ADAL auth. This is for cli < 2.30.0
+            if user_type == 'user':
+                identity_data = {}
+                identity_data['refreshToken'] = creds
+                identity_uri = f'https://localhost:{api_server_port}/identity/rt'
+
+                # Needed to prevent skip tls warning from printing to the console
+                original_stderr = sys.stderr
+                f = open(os.devnull, 'w')
+                sys.stderr = f
+
+                make_api_call_with_retries(identity_uri, identity_data, "post", False, consts.Post_RefreshToken_Fault_Type,
+                                           'Unable to post refresh token details to clientproxy',
+                                           "Failed to pass refresh token details to proxy.", clientproxy_process)
+                sys.stderr = original_stderr
+
+    if utils.is_cli_using_msal_auth():  # jwt token approach if cli is using MSAL. This is for cli >= 2.30.0
+        kid = fetch_pop_publickey_kid(api_server_port, clientproxy_process)
+        post_at_response = fetch_and_post_at_to_csp(cmd, api_server_port, tenantId, kid, clientproxy_process)
+        if post_at_response.status_code == 500 and "public key expired" in post_at_response.text:  # pop public key must have been rotated
```

Comment:
For my clarity: Will this code executed as we are calling `close_subprocess_and_raise_cli_error` in the end of `make_api_call_with_retries`?

Reply:
close_subprocess_and_raise_cli_error will happen only when there is an exception while making the CSP api call. That is, if at all CSP process is not up and took longer to start, then the API call which we're making will not find the endpoint and it will throw an exception. In that case we'll retry and close the process after retries exhaust. But, in this case of "500" response, we will get a proper error response and it is not an exception. Which is why, this code will be executed and we will refresh the PoP public key also

Reply:
I have the tested it also by setting publickey expiry to 10 mins